### ISSUE TITLE
Confusion when dataset item count isn't evenly divisible by batch number

### ISSUE THREAD

**Issue by cnuernber** [thinktopic/cortex#14]
Currently this results in dropping the last partial batch.  This is not such a big deal when training but it can be a very large deal when running.

I am not sure this should be solved in the engine given that the implementation of the dataset can solve this but we have had two teams confused by this (albeit in their first day of using cortex) so either an error needs to happen or we need to pad the input but silently dropping the input, especially during inference is probably not the best answer.

@harold @charlesg3 @CalderBot

**Comment by charlesg3** [comment]
Dropping the last partial batch means that it doesn't make predictions on it? This seems pretty bad, I'd be okay with complaining loudly if that's easiest but padding the input seems like an even better option.

**Comment by mikera** [comment]
I prefer the idea of a "batch source" which is a slight abstraction of the traditional notion of "dataset".

A batch source is simply a procedure for assembling batches which takes a count as a parameter (and possibly also a incrementing "seed").

The batch source could have multiple different implementations:

- Taking a window of data from an underlying dataset (where window would be based on a seed) - i.e. pretty much like the current dataset behaviour
- Taking a random sample of rows from a dataset (with or without replacement)
- Calling a "generator" function, like test.check or clojure.spec
- Concatenating and/or computing a function of other batch sources (this could be used for e.g. adding random noise or linear distortions to an image)
- Pulling data from a streaming source (e.g. a core.async channel)

Worth considering a refactor to something like this? I think it would give a lot more flexibility in the long run rather than being tied to a specific dataset representation or sampling procedure.

**Comment by cnuernber** [comment]
How does that differ from the existing dataset interface which presents a sequence of batches?

**Comment by mikera** [comment]
Two main improvements I guess:
- Having an (optional?) index seed I think is better than producing a seq as it provides opportunity for random access (and repeatability without holding on to the head)
- I would avoid `batch-type` as a a parameter. That should be part of the construction of the data source.

**Comment by cnuernber** [comment]
batch-type is useful when training (so you can plot training/vs cv loss or something like that) from the training system's point of view and seed is definitely not.  Furthermore the dataset may apply augmentation/randomization if the batch type is training while for inference, holdout or cross-validation it would not do such a thing.

I think a middle ground both should perhaps be optional parameters however and this would allow the interface to survive more generalized usage than specifically training a neural network.  In other words, if the function took an opts parameter that was a (possibly empty) map then we would both be satisfied.

**Comment by harold** [comment]
Recent changes (including but not limited to: https://github.com/thinktopic/cortex/pull/136) have fixed all known issues related to the commensurability of dataset and batch sizes. New-found problems related to that should be opened as new tickets (maybe referencing this one, if appropriate).

Thanks!
